File: app/src/main/java/de/devmil/paperlaunch/view/utils/IntentSelector.java
package de.devmil.paperlaunch.view.utils;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.pm.ResolveInfo;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ExpandableListView;
import android.widget.ExpandableListView.OnChildClickListener;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TabHost;
import android.widget.TabHost.TabSpec;
import android.widget.TextView;
import android.widget.Toolbar;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

import de.devmil.paperlaunch.R;

public class IntentSelector extends Activity {
	private static final String TAG = IntentSelector.class.getSimpleName();
	private static final int CREATE_SHORTCUT_REQUEST = 1;
	
	public static String EXTRA_SHORTCUT_TEXT = "de.devmil.common.extras.SHORTCUT_TEXT";
	public static String EXTRA_STRING_SHORTCUTS = "de.devmil.common.extras.STRING_SHORTCUTS";
	public static String EXTRA_STRING_ACTIVITIES = "de.devmil.common.extras.STRING_ACTIVITIES";
	
	private LinearLayout llWait;
	private ExpandableListView lvActivities;
    private CheckBox chkShowAllActivities;
	private ExpandableListView lvShortcuts;
	private TextView txtShortcuts;

    private Toolbar mToolbar;
	
	private Thread searchThread = null;
    private Object searchThreadLockObject = new Object();
	
	private List<IntentApplicationEntry> entries = new ArrayList<IntentApplicationEntry>();
	IntentSelectorAdapter adapterActivities;
	IntentSelectorAdapter adapterShortcuts;
	
	private void addResolveInfo(ResolveInfo ri, IntentApplicationEntry.IntentType intentType, boolean addAll) {
		IntentApplicationEntry newEntry;
		try {
			if(!addAll && !ri.activityInfo.exported)
				return;
			newEntry = new IntentApplicationEntry(IntentSelector.this, ri.activityInfo.packageName);
			if(!entries.contains(newEntry)) {
				entries.add(newEntry);
			} else {
				newEntry = entries.get(entries.indexOf(newEntry));
			}
			newEntry.addResolveInfo(ri, intentType);
		} catch (NameNotFoundException e) {
			Log.e(TAG, "Error while adding a package", e);
		}
	}
	
	protected static List<IntentApplicationEntry.IntentItem> getSubList(IntentApplicationEntry entry, IntentApplicationEntry.IntentType intentType) {
		switch(intentType) {
		case Main:
		case Launcher:
			return entry.getMainActivityIntentItems();
		case Shortcut:
			return entry.getShortcutIntentItems();
		}
		return null;
	}
	
	@SuppressWarnings("deprecation")
    @Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

        synchronized (searchThreadLockObject)
        {
            if (searchThread != null) {
                searchThread.stop();
                searchThread = null;
            }
        }
		
		String shortcutText = "";
		if(getIntent().hasExtra(EXTRA_SHORTCUT_TEXT))
			shortcutText = getIntent().getStringExtra(EXTRA_SHORTCUT_TEXT);
		
		String shortcutLabel = "Shortcuts";
		if(getIntent().hasExtra(EXTRA_STRING_SHORTCUTS))
			shortcutLabel = getIntent().getStringExtra(EXTRA_STRING_SHORTCUTS);
		String activitiesLabel = "Shortcuts";
		if(getIntent().hasExtra(EXTRA_STRING_ACTIVITIES))
			activitiesLabel = getIntent().getStringExtra(EXTRA_STRING_ACTIVITIES);
		
		setContentView(R.layout.common__intentselectorview);
		
		llWait = (LinearLayout)findViewById(R.id.common__intentSelector_llWait);
//		progressWait = (ProgressBar)findViewById(R.id.intentSelector_progressWait);
		lvActivities = (ExpandableListView)findViewById(R.id.common__intentSelector_lvActivities);
        chkShowAllActivities = (CheckBox)findViewById(R.id.common__intentSelector_chkShowAllActivities);
		lvShortcuts = (ExpandableListView)findViewById(R.id.common__intentSelector_lvShortcuts);
		txtShortcuts = (TextView)findViewById(R.id.common__intentSelector_txtShortcuts);
        mToolbar = (Toolbar)findViewById(R.id.common__intentSelector_toolbar);

        setActionBar(mToolbar);
		
		txtShortcuts.setText(shortcutText);
		
		lvActivities.setOnChildClickListener(new OnChildClickListener() {
			@Override
			public boolean onChildClick(ExpandableListView parent, View v,
					int groupPosition, int childPosition, long id) {
				Intent resultIntent = new Intent(Intent.ACTION_MAIN);
				IntentApplicationEntry.IntentItem entry = (IntentApplicationEntry.IntentItem)adapterActivities.getChild(groupPosition, childPosition);
				resultIntent.setClassName(entry.getPackageName(), entry.getActivityName());
				setResultIntent(resultIntent);
				return true;
			}
		});
        chkShowAllActivities.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                startSearch();
            }
        });
		lvShortcuts.setOnChildClickListener(new OnChildClickListener() {
			@Override
			public boolean onChildClick(ExpandableListView parent, View v,
					int groupPosition, int childPosition, long id) {
				Intent shortcutIntent = new Intent(Intent.ACTION_CREATE_SHORTCUT);
				IntentApplicationEntry.IntentItem entry = (IntentApplicationEntry.IntentItem)adapterShortcuts.getChild(groupPosition, childPosition);
				shortcutIntent.setClassName(entry.getPackageName(), entry.getActivityName());
				startActivityForResult(shortcutIntent, CREATE_SHORTCUT_REQUEST);
				return false;
			}
		});
		
		TabHost tabs = (TabHost) this.findViewById(android.R.id.tabhost);
		tabs.setup();
		TabSpec tspecActivities = tabs.newTabSpec(activitiesLabel);
		tspecActivities.setIndicator(activitiesLabel);
		tspecActivities.setContent(R.id.common__intentSelector_tabActivities);
		tabs.addTab(tspecActivities);
		TabSpec tspecShortcuts = tabs.newTabSpec(shortcutLabel);
		tspecShortcuts.setIndicator(shortcutLabel);
		if(shortcutText.equals("")) {
			tspecShortcuts.setContent(R.id.common__intentSelector_tabShortcuts);
			txtShortcuts.setVisibility(View.GONE);
		}
		else {
			tspecShortcuts.setContent(R.id.common__intentSelector_tabTextShortcuts);
			lvShortcuts.setVisibility(View.GONE);
		}
		tabs.addTab(tspecShortcuts);
		
		llWait.setVisibility(View.VISIBLE);

        startSearch();
	}

    private boolean stopThread;

    private void startSearch()
    {
        Thread searchThreadCopy = null;
        synchronized (searchThreadLockObject)
        {
            if (searchThread != null) {
                searchThreadCopy = searchThread;
                stopThread = true;
            }
        }

        if(searchThreadCopy != null)
        {
            try
            {
                searchThreadCopy.join();
            }
            catch(Exception e)
            {}
        }
        stopThread = false;
        entries.clear();
        searchThread = new Thread(new Runnable() {
            @Override
            public void run() {
                IntentSelector.this.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        llWait.setVisibility(View.VISIBLE);
                    }
                });
                //this approach can kill the PackageManager if there are too many apps installed
                //				List<ResolveInfo> shortcutResolved = getPackageManager().queryIntentActivities(shortcutIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);
                //				List<ResolveInfo> mainResolved = getPackageManager().queryIntentActivities(mainIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);
                //				List<ResolveInfo> launcherResolved = getPackageManager().queryIntentActivities(launcherIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);

                PackageManager pm = getPackageManager();

                List<ResolveInfo> shortcutResolved = new ArrayList<ResolveInfo>();
                List<ResolveInfo> mainResolved = new ArrayList<ResolveInfo>();
                List<ResolveInfo> launcherResolved = new ArrayList<ResolveInfo>();

                List<ApplicationInfo> appInfos = pm.getInstalledApplications(PackageManager.GET_META_DATA);

                boolean showAll = chkShowAllActivities.isChecked();

                for(ApplicationInfo appInfo : appInfos)
                {
                    synchronized (searchThreadLockObject)
                    {
                        if(stopThread)
                            return;
                    }
                    Intent shortcutIntent = new Intent(Intent.ACTION_CREATE_SHORTCUT);
                    shortcutIntent.setPackage(appInfo.packageName);

                    List<ResolveInfo> appShortcutResolved = pm.queryIntentActivities(shortcutIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);
                    shortcutResolved.addAll(appShortcutResolved);

                    boolean addMainActivities = true;

                    if(showAll)
                    {
                        try
                        {
                            PackageInfo pi = pm.getPackageInfo(appInfo.packageName, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);

                            for(ActivityInfo ai : pi.activities)
                            {
                                ResolveInfo ri = new ResolveInfo();
                                ri.activityInfo = ai;

                                mainResolved.add(ri);
                            }

                            addMainActivities = false;
                        }
                        catch(Exception e)
                        {}
                    }

                    if(addMainActivities)
                    {
                        Intent mainIntent = new Intent(Intent.ACTION_MAIN);
                        mainIntent.setPackage(appInfo.packageName);

                        List<ResolveInfo> appMainResolved = pm.queryIntentActivities(mainIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);
                        mainResolved.addAll(appMainResolved);
                    }

                    Intent launcherIntent = new Intent(Intent.ACTION_MAIN);
                    launcherIntent.addCategory(Intent.CATEGORY_LAUNCHER);
                    launcherIntent.setPackage(appInfo.packageName);

                    List<ResolveInfo> appLauncherResolved = pm.queryIntentActivities(launcherIntent, PackageManager.GET_ACTIVITIES | PackageManager.GET_INTENT_FILTERS);
                    launcherResolved.addAll(appLauncherResolved);
                }

                for(ResolveInfo ri : shortcutResolved) {
                    addResolveInfo(ri, IntentApplicationEntry.IntentType.Shortcut, false);
                }
                for(ResolveInfo ri : mainResolved) {
                    addResolveInfo(ri, IntentApplicationEntry.IntentType.Main, showAll);
                }
                for(ResolveInfo ri : launcherResolved) {
                    addResolveInfo(ri, IntentApplicationEntry.IntentType.Launcher, false);
                }
                //sort
                IntentApplicationEntry[] entriesArray = entries.toArray(new IntentApplicationEntry[0]);
                Arrays.sort(entriesArray, new Comparator<IntentApplicationEntry>() {
					@Override
					public int compare(IntentApplicationEntry object1,
									   IntentApplicationEntry object2) {

						return object1.compareTo(object2);
					}
				});
                entries = new ArrayList<IntentApplicationEntry>(Arrays.asList(entriesArray));
                for(IntentApplicationEntry entry : entries)
                    entry.sort();
                //set
                IntentSelector.this.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        adapterActivities = new IntentSelectorAdapter(IntentSelector.this, entries, IntentApplicationEntry.IntentType.Main);
                        adapterShortcuts = new IntentSelectorAdapter(IntentSelector.this, entries, IntentApplicationEntry.IntentType.Shortcut);

                        lvActivities.setAdapter(adapterActivities);
                        lvShortcuts.setAdapter(adapterShortcuts);
                        llWait.setVisibility(View.GONE);
                    }
                });

                synchronized (searchThreadLockObject)
                {
                    searchThread = null;
                }
            }
        });
        searchThread.start();
    }
	
	private void setResultIntent(Intent intent) {
		setResult(Activity.RESULT_OK, intent);
		IntentSelector.this.finish();
	}
	
	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		if(requestCode == CREATE_SHORTCUT_REQUEST && resultCode == Activity.RESULT_OK) {
			setResultIntent(data);
		}
		super.onActivityResult(requestCode, resultCode, data);
	}
	
	class IntentSelectorAdapter extends BaseExpandableListAdapter {
		private List<IntentApplicationEntry> entries;
		private IntentApplicationEntry.IntentType intentType;
		private Context context;
		
		public IntentSelectorAdapter(Context context, List<IntentApplicationEntry> entries, IntentApplicationEntry.IntentType intentType) {
			this.context = context;
			this.intentType = intentType;
			this.entries = new ArrayList<IntentApplicationEntry>();
			for(IntentApplicationEntry entry : entries) {
				if(getSubList(entry).size() > 0)
					this.entries.add(entry);
			}
		}
		
		public List<IntentApplicationEntry.IntentItem> getSubList(IntentApplicationEntry entry) {
			return IntentSelector.getSubList(entry, intentType);
		}

		@Override
		public Object getChild(int groupPosition, int childPosition) {
			return getSubList(entries.get(groupPosition)).get(childPosition);
		}

		@Override
		public long getChildId(int groupPosition, int childPosition) {
			return groupPosition * 1000 + childPosition;
		}

		@Override
		public View getChildView(int groupPosition, int childPosition,
				boolean isLastChild, View convertView, ViewGroup parent) {
			if(convertView == null) {
				convertView = LayoutInflater.from(context).inflate(R.layout.common__intentselectoritem, null);
			}
			TextView txt = (TextView)convertView.findViewById(R.id.common__intentselectoritem_text);
			TextView txtActivityName = (TextView)convertView.findViewById(R.id.common__intentselectoritem_activityName);
			
			txt.setText(getSubList(entries.get(groupPosition)).get(childPosition).getName());
			txtActivityName.setText(getSubList(entries.get(groupPosition)).get(childPosition).getActivityName());
			return convertView;
		}

		@Override
		public int getChildrenCount(int groupPosition) {
			return getSubList(entries.get(groupPosition)).size();
		}

		@Override
		public Object getGroup(int groupPosition) {
			return entries.get(groupPosition);
		}

		@Override
		public int getGroupCount() {
			return entries.size();
		}

		@Override
		public long getGroupId(int groupPosition) {
			return groupPosition * 1000;
		}

		@Override
		public View getGroupView(int groupPosition, boolean isExpanded,
				View convertView, ViewGroup parent) {
			if(convertView == null) {
				convertView = LayoutInflater.from(context).inflate(R.layout.common__intentselectorgroup, null);
			}
			ImageView img = (ImageView)convertView.findViewById(R.id.common__intentselectorgroup_img);
			TextView txt = (TextView)convertView.findViewById(R.id.common__intentselectorgroup_text);
			
			txt.setText(entries.get(groupPosition).getName());
			Drawable appIcon = entries.get(groupPosition).getAppIcon();
			if(appIcon != null) {
				img.setImageDrawable(entries.get(groupPosition).getAppIcon());
				img.setVisibility(View.VISIBLE);
			} else {
				img.setVisibility(View.INVISIBLE);
			}
			return convertView;
		}

		@Override
		public boolean hasStableIds() {
			return false;
		}

		@Override
		public boolean isChildSelectable(int groupPosition, int childPosition) {
			return true;
		}
	}
}

File: app/src/main/java/de/devmil/paperlaunch/view/utils/IntentSelector.kt
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.view.utils

import android.app.Activity
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.content.pm.ResolveInfo
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import de.devmil.paperlaunch.R
import java.lang.ref.WeakReference
import java.util.*

class IntentSelector : AppCompatActivity(), SearchView.OnQueryTextListener {

    private var llWait: LinearLayout? = null
    private var lvActivities: ExpandableListView? = null
    private var chkShowAllActivities: CheckBox? = null
    private var lvShortcuts: ExpandableListView? = null
    private var txtShortcuts: TextView? = null

    private var toolbar: Toolbar? = null

    class SearchTask
    constructor(intentSelector : IntentSelector) : AsyncTask<Unit, Int, Unit>() {

        private val entries = mutableListOf<IntentApplicationEntry>()

        var isAnotherSearchRunning: Boolean = false
        var isObsolete: Boolean = false

        @Deprecated("Deprecated in Java")
        override fun doInBackground(vararg params: Unit?) {
            intentSelectorRef.get()?.let {
                val pm = it.packageManager

                val shortcutResolved = ArrayList<ResolveInfo>()
                val mainResolved = ArrayList<ResolveInfo>()
                val launcherResolved = ArrayList<ResolveInfo>()

                val showAll = it.chkShowAllActivities!!.isChecked

                // Instead of getInstalledApplications, we query intents directly for better performance

                // 1. Launcher Activities (Always needed)
                val launcherIntent = Intent(Intent.ACTION_MAIN)
                launcherIntent.addCategory(Intent.CATEGORY_LAUNCHER)
                try {
                    launcherResolved.addAll(pm.queryIntentActivities(launcherIntent, 0))
                } catch (e: Exception) {
                    Log.e(TAG, "Error querying launcher activities", e)
                }

                if (isCancelled || isObsolete) return

                // 2. Shortcuts (Always needed)
                val shortcutIntent = Intent(Intent.ACTION_CREATE_SHORTCUT)
                try {
                    shortcutResolved.addAll(pm.queryIntentActivities(shortcutIntent, 0))
                } catch (e: Exception) {
                    Log.e(TAG, "Error querying shortcut activities", e)
                }

                if (isCancelled || isObsolete) return

                // 3. Main Activities (Only if showAll is true)
                // Note: Querying ACTION_MAIN without category can be huge, but if user requests it...
                // We can optimize by iterating launcherResolved/shortcutResolved to find packages?
                // Or just query only if checked.
                if (showAll) {
                    val mainIntent = Intent(Intent.ACTION_MAIN)
                    try {
                        mainResolved.addAll(pm.queryIntentActivities(mainIntent, 0))
                    } catch (e: Exception) {
                        Log.e(TAG, "Error querying main activities", e)
                    }
                }

                // Process Launcher Activities
                for (ri in launcherResolved) {
                    if (isCancelled || isObsolete) return
                    addResolveInfo(ri, IntentApplicationEntry.IntentType.Launcher, false, entries)
                }

                // Process Shortcut Activities
                for (ri in shortcutResolved) {
                     if (isCancelled || isObsolete) return
                     addResolveInfo(ri, IntentApplicationEntry.IntentType.Shortcut, false, entries)
                }

                // Process Main Activities
                if (showAll) {
                    for (ri in mainResolved) {
                        if (isCancelled || isObsolete) return
                        addResolveInfo(ri, IntentApplicationEntry.IntentType.Main, true, entries)
                    }
                }

                //sort
                val comparator = Comparator<IntentApplicationEntry> { object1, object2 -> object1.compareTo(object2) }
                entries.sortWith(comparator)
                entries.forEach { if(!isCancelled && !isObsolete) it.sort() }
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onPreExecute() {
            super.onPreExecute()
            val localIntentSelector = intentSelectorRef.get()
            localIntentSelector?.runOnUiThread { localIntentSelector.llWait!!.visibility = View.VISIBLE }
        }

        @Deprecated("Deprecated in Java")
        override fun onPostExecute(result: Unit?) {
            super.onPostExecute(result)
            val localIntentSelector = intentSelectorRef.get()
            localIntentSelector?.runOnUiThread {
                if(!isCancelled && !isObsolete) {
                    localIntentSelector.adapterActivities = IntentSelectorAdapter(localIntentSelector, entries, IntentApplicationEntry.IntentType.Main)
                    localIntentSelector.adapterShortcuts = IntentSelectorAdapter(localIntentSelector, entries, IntentApplicationEntry.IntentType.Shortcut)

                    localIntentSelector.lvActivities!!.setAdapter(localIntentSelector.adapterActivities)
                    localIntentSelector.lvShortcuts!!.setAdapter(localIntentSelector.adapterShortcuts)

                    localIntentSelector.adapterActivities?.filter(localIntentSelector.mCurrentQuery)
                    localIntentSelector.adapterShortcuts?.filter(localIntentSelector.mCurrentQuery)
                }
                if(!isAnotherSearchRunning) {
                    localIntentSelector.llWait!!.visibility = View.GONE
                }
            }
        }

        @Deprecated("Deprecated in Java")
        override fun onCancelled() {
            super.onCancelled()
            val localIntentSelector = intentSelectorRef.get()
            localIntentSelector?.runOnUiThread {
                if(!isAnotherSearchRunning) {
                    localIntentSelector.llWait!!.visibility = View.GONE
                }
            }
        }

        private fun addResolveInfo(ri: ResolveInfo, intentType: IntentApplicationEntry.IntentType, addAll: Boolean, entries : MutableList<IntentApplicationEntry>) {
            intentSelectorRef.get()?.let {
                try {
                    if (!addAll && !ri.activityInfo.exported)
                        return
                    var newEntry = IntentApplicationEntry(it, ri.activityInfo.packageName)
                    if (!entries.contains(newEntry)) {
                        entries.add(newEntry)
                    } else {
                        newEntry = entries[entries.indexOf(newEntry)]
                    }
                    newEntry.addResolveInfo(ri, intentType)
                } catch (e: NameNotFoundException) {
                    Log.e(TAG, "Error while adding a package", e)
                }
            }
        }

        private var intentSelectorRef: WeakReference<IntentSelector> = WeakReference(intentSelector)
    }
    private var mSearchTask : SearchTask? = null
    private var adapterActivities: IntentSelectorAdapter? = null
    private var adapterShortcuts: IntentSelectorAdapter? = null

    private var fabDone: android.support.design.widget.FloatingActionButton? = null
    private var allowMultiSelect: Boolean = false
    private val selectedIntents = ArrayList<Intent>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        allowMultiSelect = intent.getBooleanExtra(EXTRA_ALLOW_MULTI_SELECT, false)

        if(mSearchTask != null) {
            mSearchTask!!.cancel(true)
            mSearchTask = null
        }

        var shortcutText = ""
        if (intent.hasExtra(EXTRA_SHORTCUT_TEXT))
            shortcutText = intent.getStringExtra(EXTRA_SHORTCUT_TEXT)!!

        var shortcutLabel = "Shortcuts"
        if (intent.hasExtra(EXTRA_STRING_SHORTCUTS))
            shortcutLabel = intent.getStringExtra(EXTRA_STRING_SHORTCUTS)!!
        var activitiesLabel = "Shortcuts"
        if (intent.hasExtra(EXTRA_STRING_ACTIVITIES))
            activitiesLabel = intent.getStringExtra(EXTRA_STRING_ACTIVITIES)!!

        setContentView(R.layout.common__intentselectorview)

        llWait = findViewById(R.id.common__intentSelector_llWait)
        //		progressWait = (ProgressBar)findViewById(R.id.intentSelector_progressWait);
        lvActivities = findViewById(R.id.common__intentSelector_lvActivities)
        chkShowAllActivities = findViewById(R.id.common__intentSelector_chkShowAllActivities)
        lvShortcuts = findViewById(R.id.common__intentSelector_lvShortcuts)
        txtShortcuts = findViewById(R.id.common__intentSelector_txtShortcuts)
        toolbar = findViewById(R.id.common__intentSelector_toolbar)

        setSupportActionBar(toolbar)

        txtShortcuts!!.text = shortcutText

        lvActivities!!.setOnChildClickListener { _, _, groupPosition, childPosition, _ ->
            val entry = adapterActivities!!.getChild(groupPosition, childPosition) as IntentApplicationEntry.IntentItem
            val resultIntent = Intent(Intent.ACTION_MAIN)
            resultIntent.setClassName(entry.packageName, entry.activityName)

            if (allowMultiSelect) {
                toggleSelection(resultIntent)
                adapterActivities!!.notifyDataSetChanged()
                true
            } else {
                setResultIntent(resultIntent)
                true
            }
        }
        chkShowAllActivities!!.setOnCheckedChangeListener { _, _ -> startSearch() }
        lvShortcuts!!.setOnChildClickListener { _, _, groupPosition, childPosition, _ ->
            val entry = adapterShortcuts!!.getChild(groupPosition, childPosition) as IntentApplicationEntry.IntentItem
            val shortcutIntent = Intent(Intent.ACTION_CREATE_SHORTCUT)
            shortcutIntent.setClassName(entry.packageName, entry.activityName)
            startActivityForResult(shortcutIntent, CREATE_SHORTCUT_REQUEST)
            false
        }

        fabDone = findViewById(R.id.common__intentSelector_fab)
        if (allowMultiSelect) {
            fabDone!!.show()
            fabDone!!.setOnClickListener {
                val resultIntent = Intent()
                resultIntent.putParcelableArrayListExtra(EXTRA_RESULT_INTENTS, selectedIntents)
                setResult(Activity.RESULT_OK, resultIntent)
                finish()
            }
            // Keep tabs but maybe hide shortcuts for multiselect if not supported?
            // Assuming shortcuts are not supported in multi-select for now or handled same way?
            // Existing logic launches helper activity for shortcuts. This breaks multi-select flow unless handled.
            // For now, let's just make Multi-Select work for Apps which is the performance killer.
        } else {
            fabDone!!.hide()
        }

        val tabs = this.findViewById<TabHost>(android.R.id.tabhost)
        tabs.setup()
        val tspecActivities = tabs.newTabSpec(activitiesLabel)
        tspecActivities.setIndicator(activitiesLabel)
        tspecActivities.setContent(R.id.common__intentSelector_tabActivities)
        tabs.addTab(tspecActivities)
        val tspecShortcuts = tabs.newTabSpec(shortcutLabel)
        tspecShortcuts.setIndicator(shortcutLabel)
        if (shortcutText == "") {
            tspecShortcuts.setContent(R.id.common__intentSelector_tabShortcuts)
            txtShortcuts!!.visibility = View.GONE
        } else {
            tspecShortcuts.setContent(R.id.common__intentSelector_tabTextShortcuts)
            lvShortcuts!!.visibility = View.GONE
        }
        tabs.addTab(tspecShortcuts)

        llWait!!.visibility = View.VISIBLE

        startSearch()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_intent_selector, menu)
        val searchItem = menu.findItem(R.id.action_search)
        val searchView = searchItem.actionView as SearchView
        searchView.setOnQueryTextListener(this)
        return true
    }

    override fun onQueryTextSubmit(query: String?): Boolean {
        return false
    }

    private var mCurrentQuery: String? = null

    override fun onQueryTextChange(newText: String?): Boolean {
        mCurrentQuery = newText
        adapterActivities?.filter(newText)
        adapterShortcuts?.filter(newText)
        return true
    }

    private fun startSearch() {
        if (mSearchTask != null) {
            mSearchTask!!.isAnotherSearchRunning = true
            mSearchTask!!.isObsolete = true
            mSearchTask!!.cancel(true)
            mSearchTask = null
        }

        mSearchTask = SearchTask(this)
        mSearchTask!!.execute()
    }

    private fun setResultIntent(intent: Intent) {
        setResult(Activity.RESULT_OK, intent)
        this@IntentSelector.finish()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if(data == null) {
            return
        }
        if (requestCode == CREATE_SHORTCUT_REQUEST && resultCode == Activity.RESULT_OK) {
            setResultIntent(data)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun toggleSelection(intent: Intent) {
        // Simple check based on component name
        val component = intent.component ?: return
        val existing = selectedIntents.find { it.component == component }
        if (existing != null) {
            selectedIntents.remove(existing)
        } else {
            selectedIntents.add(intent)
        }
        updateTitleInternal()
    }

    private fun updateTitleInternal() {
        if (!allowMultiSelect) return
        if (selectedIntents.size > 0) {
            title = getString(R.string.activity_intentselector_title_count, selectedIntents.size)
        } else {
            title = getString(R.string.activity_intentselector_label)
        }
    }

    private fun isSelected(intent: Intent): Boolean {
        val component = intent.component ?: return false
        return selectedIntents.any { it.component == component }
    }

    internal class IntentSelectorAdapter(private val context: Context, entriesList: List<IntentApplicationEntry>, private val intentType: IntentApplicationEntry.IntentType) : BaseExpandableListAdapter() {
        private var entries: MutableList<IntentApplicationEntry>
        private val originalEntries: List<IntentApplicationEntry>

        init {
            this.originalEntries = entriesList
                    .filter { getSubList(it).isNotEmpty() }
            this.entries = this.originalEntries.toMutableList()
        }

        fun filter(query: String?) {
            entries.clear()
            if (query.isNullOrEmpty()) {
                entries.addAll(originalEntries)
            } else {
                val q = query.lowercase(Locale.getDefault())
                for (entry in originalEntries) {
                    var matches = false
                    if (entry.nameLowercase.contains(q)) {
                        matches = true
                    } else {
                        val subList = getSubList(entry)
                        if (subList.any { it.displayNameLowercase.contains(q) }) {
                            matches = true
                        }
                    }

                    if (matches) {
                        entries.add(entry)
                    }
                }
            }
            notifyDataSetChanged()
        }

        fun getSubList(entry: IntentApplicationEntry): List<IntentApplicationEntry.IntentItem> {
            return IntentSelector.getSubList(entry, intentType)
        }

        override fun getChild(groupPosition: Int, childPosition: Int): Any {
            return getSubList(entries[groupPosition])[childPosition]
        }

        override fun getChildId(groupPosition: Int, childPosition: Int): Long {
            return (groupPosition * 1000 + childPosition).toLong()
        }

        override fun getChildView(groupPosition: Int, childPosition: Int,
                                  isLastChild: Boolean, convertView: View?, parent: ViewGroup): View {
            var effectiveConvertView = convertView
            if (effectiveConvertView == null) {
                effectiveConvertView = LayoutInflater.from(context).inflate(R.layout.common__intentselectoritem, parent, false)
            }
            val txt = effectiveConvertView!!.findViewById<TextView>(R.id.common__intentselectoritem_text)
            val txtActivityName = effectiveConvertView.findViewById<TextView>(R.id.common__intentselectoritem_activityName)
            val checkbox = effectiveConvertView.findViewById<CheckBox>(R.id.common__intentselectoritem_checkbox)

            val item = getSubList(entries[groupPosition])[childPosition]
            txt.text = item.displayName
            txtActivityName.text = item.activityName

            if ((context as IntentSelector).allowMultiSelect && intentType == IntentApplicationEntry.IntentType.Main) {
                checkbox.visibility = View.VISIBLE
                val testIntent = Intent(Intent.ACTION_MAIN)
                testIntent.setClassName(item.packageName, item.activityName)
                checkbox.isChecked = context.isSelected(testIntent)
            } else {
                checkbox.visibility = View.GONE
            }

            return effectiveConvertView
        }

        override fun getChildrenCount(groupPosition: Int): Int {
            return getSubList(entries[groupPosition]).size
        }

        override fun getGroup(groupPosition: Int): Any {
            return entries[groupPosition]
        }

        override fun getGroupCount(): Int {
            return entries.size
        }

        override fun getGroupId(groupPosition: Int): Long {
            return (groupPosition * 1000).toLong()
        }

        override fun getGroupView(groupPosition: Int, isExpanded: Boolean,
                                  convertView: View?, parent: ViewGroup): View {
            var effectiveConvertView = convertView
            if (effectiveConvertView == null) {
                effectiveConvertView = LayoutInflater.from(context).inflate(R.layout.common__intentselectorgroup, parent, false)
            }
            val img = effectiveConvertView!!.findViewById<ImageView>(R.id.common__intentselectorgroup_img)
            val txt = effectiveConvertView.findViewById<TextView>(R.id.common__intentselectorgroup_text)

            txt.text = entries[groupPosition].name
            val appIcon = entries[groupPosition].getAppIcon()
            if (appIcon != null) {
                img.setImageDrawable(entries[groupPosition].getAppIcon())
                img.visibility = View.VISIBLE
            } else {
                img.visibility = View.INVISIBLE
            }
            return effectiveConvertView
        }

        override fun hasStableIds(): Boolean {
            return false
        }

        override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
            return true
        }
    }

    companion object {
        private val TAG = IntentSelector::class.java.simpleName
        private val CREATE_SHORTCUT_REQUEST = 1

        var EXTRA_SHORTCUT_TEXT = "de.devmil.common.extras.SHORTCUT_TEXT"
        var EXTRA_STRING_SHORTCUTS = "de.devmil.common.extras.STRING_SHORTCUTS"
        var EXTRA_STRING_ACTIVITIES = "de.devmil.common.extras.STRING_ACTIVITIES"
        var EXTRA_ALLOW_MULTI_SELECT = "de.devmil.common.extras.ALLOW_MULTI_SELECT"
        var EXTRA_RESULT_INTENTS = "de.devmil.common.extras.RESULT_INTENTS"

        private fun getSubList(entry: IntentApplicationEntry, intentType: IntentApplicationEntry.IntentType): List<IntentApplicationEntry.IntentItem> {
            when (intentType) {
                IntentApplicationEntry.IntentType.Main, IntentApplicationEntry.IntentType.Launcher -> return entry.getMainActivityIntentItems()
                IntentApplicationEntry.IntentType.Shortcut -> return entry.getShortcutIntentItems()
            }
        }
    }
}
